### ISSUE TITLE
--save-json doesn't work properly in val.py

### ISSUE THREAD

**Issue by melika-sce** [ultralytics/yolov5#12104]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Validation

### Bug

Hi
I need the prediction result of trained yolov5 model in coco format, as I searched in the code, by using --save-json in val.py I can get the predictions in json format and then I only have to change the image ids with image names so that they can be useful by any coco evaluators. but when I run the code with the below command, it saves different number of predictions each time
`val.py --data VisDrone.yaml --weights weights/best.pt --batch-size 16 --imgsz 1024 --save-json`

note that VisDrone dataset has 548 images in validation folder but the numbers saved from running the code is 528, 534, ... and never the right number(548)!

I test this command on both my machine and google colab and with different trained models (yolov5n, yolov5s) but they both encounter the same problem.

I would appreciate any suggestions or help!

### Environment

-YOLO: YOLOv5  v7.0-215-ga6659d0 Python-3.8.17 torch-2.0.1 CUDA:0 (NVIDIA GeForce 940MX, 2048MiB)
-OS: Windows 10
-python: 3.8

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@melika-sce hi,

Thank you for bringing this issue to our attention. The `--save-json` option in `val.py` is designed to save predictions in JSON format for further evaluation. We apologize for the inconvenience you are facing.

As for the discrepancy in the number of predictions saved, it could be due to several reasons such as skipped images or errors encountered during inference. To investigate this further, we would need a minimal reproducible example that includes both the dataset and the trained model. With that information, we can try to reproduce the issue and provide a more accurate solution or workaround.

In the meantime, you can manually convert the image IDs to image names in the saved JSON file using a simple script. This will allow you to use the COCO evaluators on the converted JSON file.

We appreciate your understanding and patience. Please let us know if you have any further questions or need additional assistance.

Best regards!

**Comment by melika-sce** [comment]
Thanks for the response, here is the command I used to train my model on VisDrone dataset:

`python train.py --data VisDrone.yaml --epochs 50 --imgsz 1024 --weights yolov5s.pt  --batch-size 16`

the trained model can be accessed via this link : 

https://drive.google.com/drive/folders/1oHuLGTUQWFjXCczOvxbvlQrnIaeQNTWA?usp=sharing


and here is the command I used to evaluate my model and save the predictions in JSON format:

`val.py --data VisDrone.yaml --weights weights/best.pt --batch-size 16 --imgsz 1024 --save-json`

**Comment by glenn-jocher** [comment]
@melika-sce thanks for sharing the details of your training and evaluation commands, as well as providing the link to your trained model.

To further investigate the issue with the `--save-json` option in `val.py` not saving the correct number of predictions, we would need access to the VisDrone dataset you used for training and evaluation.

Please upload a small subset of the VisDrone dataset, including a few images from the validation set, to a publicly accessible location (such as Google Drive, Dropbox, or GitHub) so that we can reproduce the issue on our end.

Once you provide access to the dataset, we will be able to run the evaluation command using your trained model and investigate the discrepancy in the number of predictions saved.

We appreciate your cooperation and look forward to resolving this issue for you.

**Comment by melika-sce** [comment]
This is the link for VisDrone validation dataset that I used (containing 548 images and annotations):

https://drive.google.com/drive/folders/1DsEzMk0I8YCUZS1vFXdGgsgj4vYzlhx3?usp=sharing

Thanks for your help

**Comment by glenn-jocher** [comment]
Thank you for providing the VisDrone validation dataset, @melika-sce.

We apologize for the inconvenience you are experiencing with the `--save-json` option in `val.py`. We understand the importance of having accurate predictions saved in JSON format for further evaluation.

We will investigate the issue further using the dataset you provided. Our team will run the evaluation command with your trained model and the VisDrone validation dataset to reproduce the problem and find a solution.

We appreciate your patience and cooperation. We will do our best to resolve this issue and provide a solution as soon as possible.

Thank you for your understanding.

Glenn Jocher, YOLOv5 Team
